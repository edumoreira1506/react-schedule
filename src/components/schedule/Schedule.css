.Schedule {
    text-align: center;
    width: 60%;
}
    .Schedule-header {
        font-size: 40px;
        font-style: italic;
        font-weight: 400;
        margin: 0 0 20px 0;
    }
        .Schedule-header-container {
            max-width: 80%;
            margin: 0 auto;
        }
            .Schedule-header-container-form {
                height: 60px;
                background: #ccc;
                border-radius: 3px;
                border: solid 1px #00000059;
            }
    .Schedule-border {
        width: 65%;
        height: 2px;
        background: #ccc;
        margin: 20px 0 20px 0;
        display: inline-block;
        text-align: center;
    }
    .Schedule-container-table{
        height: 50vh;
        overflow: hidden auto;
        margin-bottom: 16px;
    }
        .Schedule-table {
            width: 80%;
            height: auto;
            max-height: 35vh;
            margin: 0 auto;
            background: #ccc;
            border-radius: 3px;
            border: solid 1px #00000059;
            color: #000;
            font-weight: 500;
            text-transform: uppercase;
        }
            .Schedule-table-thead{
                background: #f5f5f5;
                color: #ccc;
            }
            .Schedule-table-thead:hover{
                background: #ccc;
                color: #f5f5f5;
            }