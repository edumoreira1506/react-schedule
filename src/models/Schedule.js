const validate = (fields) => {
  for(let i in fields){
    if(!fields[i]){
      return {
        ok: false,
        field: i
      }
    }
  }

  return {
    ok: true
  }
}

export { validate }