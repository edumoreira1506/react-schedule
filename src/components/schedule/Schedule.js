import React, { Component } from 'react'
import DefaultTable from './../table/DefaultTable'
import DefaultForm from './../form/DefaultForm'

import { validate } from './../../models/Schedule'

import {
  brazilianDate
} from './../../helpers/date'

import './Schedule.css'

class Schedule extends Component{

  constructor(props){
    super(props)

    this.state = {
      items: [],
      time: '',
      type: 'Run',
      date: '',
      studiedHours: 0,
    }
  }

  handleChange = (e) => {
    this.setState({
      [e.target.name]: e.target.value
    })
  }
    
  removeItem = (index) => {
    const items = this.state.items

    let studiedHours = this.state.studiedHours 
    studiedHours -= parseInt(items[index][0])

    items.splice(index, 1)

    this.setState({ items, studiedHours });
  }

  addItem = (e) => {
    e.preventDefault()

    const requiredFields = [this.state.time, this.state.date, this.state.type]
    const validated = validate(requiredFields)

    if(!validated.ok){
      alert(`For add new item, no one field can be empty.`)
    }else{
      const items = [
        ...this.state.items,
        [
           this.state.time,
           this.state.type,
           brazilianDate(this.state.date),
        ]
      ]

      let studiedHours = this.state.studiedHours
      studiedHours += parseInt(this.state.time)

      this.setState({ 
        items, 
        studiedHours,
        time: '',
        type: 'Run',
        date: ''
      });
    }
  }

  render() {
    const inputs = [
      {
        onChange: this.handleChange,
        value: this.state.time,
        name: 'time',
        type: 'number',
        placeholder: 'Time event'
      },
      {
        onChange: this.handleChange,
        name: 'date',
        value: this.state.date,
        type: 'date',
      }
    ]

    const selects = [
      {
        onChange: this.handleChange,
        name: 'type',
        value: this.state.type,
        options: ['Run', 'Speed']
      }
    ]

    const itemsHead = ['Time', 'Type', 'Date', 'Actions']

    return (
      <div className="Schedule">
        <h1 className="Schedule-header">{this.props.title}</h1>
        <DefaultForm selects={selects} inputs={inputs} onSubmit={this.addItem} />
        <div className="Schedule-border"></div>
        <div className="Schedule-container-table">
          <DefaultTable onClick={this.removeItem} itemsHead={itemsHead} items={this.state.items} />
        </div>
        <div>
          <h3>{this.state.studiedHours} hours</h3>
        </div>
      </div>
    )
  }
}

export default Schedule
