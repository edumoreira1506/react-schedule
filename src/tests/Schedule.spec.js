import React from 'react'
import { create } from 'react-test-renderer'
import Schedule from './../components/schedule/Schedule'

const title = 'TITLE HERE'
const emptyArray = new Array()
const emptyString = ''
const zero = 0

const makeSchedule = (title) => {
    return create(<Schedule title={title} />)
}

describe('Schedule component', () => {
    const schedule = makeSchedule(title)
    const instance = schedule.getInstance()

    test('Matches the snapshot', () => {
        expect(schedule.toJSON()).toMatchSnapshot()
    })

    test('Title of schedule', () => {
        expect(instance.props.title).toBe(title)
    })

    test('Empty itens on new instace', () => {
        expect(instance.state.items).toStrictEqual(emptyArray)
        expect(instance.state.time).toBe(emptyString)
        expect(instance.state.type).toBe(emptyString)
        expect(instance.state.date).toBe(emptyString)
        expect(instance.state.studiedHours).toBe(zero)
    })

})