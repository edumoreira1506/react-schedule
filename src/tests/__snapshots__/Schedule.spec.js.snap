// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schedule component Matches the snapshot 1`] = `
<div
  className="Schedule"
>
  <h1
    className="Schedule-header"
  >
    TITLE HERE
  </h1>
  <div
    className="Schedule-header-container"
  >
    <form
      className="Schedule-header-container-form"
      onSubmit={[Function]}
    >
      <input
        className="input-group"
        onChange={[Function]}
        placeholder="Time event"
        type="number"
        value=""
      />
      <input
        className="input-group"
        onChange={[Function]}
        type="date"
        value=""
      />
      <select
        className="input-group"
        onChange={[Function]}
        value=""
      >
        <option
          value=""
        >
          Selecionar
        </option>
        <option
          value="Run"
        >
          Run
        </option>
        <option
          value="Speed"
        >
          Speed
        </option>
      </select>
      <button
        className="input-group transition"
        type="submit"
      >
        Add
      </button>
    </form>
  </div>
  <div
    className="Schedule-border"
  />
  <div
    className="Schedule-container-table"
  >
    <table
      className="Schedule-table"
    >
      <thead
        className="Schedule-table-thead transition"
      >
        <tr>
          <td
            className="Cell-row"
          >
            Time
          </td>
          <td
            className="Cell-row"
          >
            Type
          </td>
          <td
            className="Cell-row"
          >
            Date
          </td>
          <td
            className="Cell-row"
          >
            Actions
          </td>
        </tr>
      </thead>
      <tbody
        className="Schedule-table-tbody"
      />
    </table>
  </div>
  <div>
    <h3>
      0
       hours
    </h3>
  </div>
</div>
`;
